### ISSUE TITLE
Required fields no longer working

### ISSUE THREAD

**Issue by CaptJiggly** [themeum/kirki#1837]
### Issue description:
Required fields don't seem to be working. I'm not sure what's causing it. Kirki_helper::compare_values and field-dependencies.js both seem okay.

In your post (https://github.com/aristath/kirki/issues/1709) you mentioned something about themes using

`transport => 'auto'`

But, the theme I am working on uses the default refresh. We don't even override the value.

Can you check on this issue, or perhaps I'm missing something.

### Version used:
(Did you try using the develop branch from github? There's a chance your issue has already been adressed there)
postMessage-refractor

### Using theme_mods or options?
theme_mods

### Code to reproduce the issue (config + field(s))
Any required fields.

**Comment by aristath** [comment]
@CaptJiggly I'm unable to replicate this issue... I just tested the `develop` and `postMessage-refactor` branches, everything seems to be working fine on my end.
Could you post an example that fails for you?

**Comment by CaptJiggly** [comment]
I had some time to do some extra digging and ran into this on the JS side of things:

![fail](https://user-images.githubusercontent.com/14947514/37567713-d4a22458-2a98-11e8-9aef-25d1c8a53617.png)

The code for those three:

```php
Kirki::add_section('header_bottom_setup_section', array(
    'title' => esc_html__('Setup', 'translator'),
    'panel' => 'header_bottom_panel'
));

Kirki::add_field($opt_name, array(
    'label' => esc_html__('Header bottom', 'translator'),
    'section' => 'header_bottom_setup_section',
    'settings' => 'header_bottom_visibility',
    'type' => 'multicheck',
    'choices' => array(
        'mobile' => esc_html__('Mobile', 'translator'),
        'tablet' => esc_html__('Tablet', 'translator'),                
        'desktop' => esc_html__('Desktop', 'translator')
    ),
    'default' => array()
));

Kirki::add_field($opt_name, array(
    'label' => esc_html__('Header bottom background style', 'translator'),
    'section' => 'header_bottom_setup_section',
    'settings' => 'header_bottom_bg_style',
    'type' => 'radio-buttonset',
    'required' => array(
        array(
			'setting' => 'header_bottom_visibility', 
			'operator' => 'contains', 
			'value' => array('mobile', 'tablet', 'desktop')
        )
    ),
    'choices' => array(
        'classic' => esc_html__('Classic', 'translator'),
        'gradient' => esc_html__('Gradient', 'translator')
    ),
    'default' => 'classic'
));

Kirki::add_field($opt_name, array(
    'label' => esc_html__('Header bottom background gradient first color location', 'translator'),
    'section' => 'header_bottom_setup_section',
    'settings' => 'header_bottom_bg_gradient_from_color_location',
    'type' => 'slider',
    'min' => 0,
    'step' => 1,
    'max' => 50,
    'required' => array(
        array('header_bottom_visibility', 'contains', array('mobile', 'tablet', 'desktop')),
        array('header_bottom_bg_style', '==', 'gradient')
    ),
    'default' => 0
));
```
They were in split sections, but I put them in the same just for ease of testing.

Any insight on what might be causing that?

Its happening with a couple more fields from what I can tell, but this was the first one having issues.

I have the full error log if you need it, but the first three lines repeat until JS gives up.

**Comment by CaptJiggly** [comment]
@aristath Any possibly idea on what could be causing that?

**Comment by CaptJiggly** [comment]
I found the issue.

So the problem we were having was in `field-dependencies.js`, it is assuming that the required statement will be a multi-dimensional array.

Eg:

`'required' => array(
array('setting', '=', 'something')
)`

But, in some cases, were were not 

`required => array('setting', '=', 'something')`

and this block:

`if ( 'undefined' !== typeof requirement[0] && 'undefined' === typeof requirement.setting ) {`
	`nestedItems = [];`

	// Loop sub-requirements.
	for ( i = 0; i < requirement.length; i++ ) {
		nestedItems.push( self.checkCondition( requirement[ i ], control, isOption, childRelation ) );
	}

	// OR relation. Check that true is part of the array.
	if ( 'OR' === childRelation ) {
		return ( -1 !== nestedItems.indexOf( true ) );
	}

	// AND relation. Check that false is not part of the array.
	return ( -1 === nestedItems.indexOf( false ) );
}

Since it was assuming a multi dimensional, it was trying to take `setting` and attempt to check the condition of each character... Hench the stack overflow.

So what I did was add Array.isArray

`
if ( Array.isArray(requirement) && 'undefined' !== typeof requirement[0] && 'undefined' === typeof requirement.setting ) {
`

And that took care of our issue and handles both single and multi dimensional arrays for requirements.

**Comment by aristath** [comment]
@CaptJiggly I'm sorry for the delay here, time is very limited lately and I didn't have any time to check this. 
Is this still an issue? 
If yes, could you please submit a pull-request with your fix? As soon as this one gets fixed I think v3.0.26 can be released on wordpress.org

**Comment by DannyCooper** [comment]
This seems like an implementation error and the PR never came. So closing now.
